reactive/kotlinx-coroutines-reactor/src/Mono.kt

```diff
@@ -64,18 +59,24 @@ private class MonoCoroutine<in T>(
     parentContext: CoroutineContext,
     private val sink: MonoSink<T>
 ) : AbstractCoroutine<T>(parentContext, true), Disposable {
-    var disposed = false
+    @Volatile
+    private var disposed = false
 
     override fun onCompleted(value: T) {
-        if (!disposed) {
-            if (value == null) sink.success() else sink.success(value)
-        }
+        if (value == null) sink.success() else sink.success(value)
```

Comment:
Removal of `if (!disposed) ...` check around this emission and also inside `onCancelled` function below makes behavior of `mono` builder inconsistent with the behavior of `flux` , which does not call any methods on subscriber after it was disposed. 

Note, that the reactive streams specs does not actually prohibit calling any `onXxx` methods in this case. It just says in Rule 8: "If a Subscription is cancelled its Subscriber MUST eventually stop being signaled."

Same question applies to `RxCompletableCoroutine` changes.

Reply:
It is, in fact, a really small inconsistency.
To make this race happen, a subscriber should manually call `dispose` on its subscription while coroutine is in its completion sequence (CAS in JobSupport is successful, so `job.cancel` doesn't affect anything). In this case, `if (!disposed)` makes a really small race window even smaller, but does not prevent it anyway. 

Adding `if (!isDisposed)` adds a cognitive load for readers, because it's "check-and-act" on some volatile state that is hard to reason about (yeah, we can leave a comment, but what's the point?), but does not change the semantics at all (because "race" is still possible).

Similar reasoning can be applied to `onCancelled` and Rx.